src/fast-instantiator.ts

```diff
@@ -0,0 +1,205 @@
+  instance: Instance<T>,
+  snapshot: SnapshotIn<T>,
+  context: InstantiateContext
+) => void;
+
+/**
+ * Compiles a fast function for taking snapshots and turning them into instances of a class model.
+ **/
+export const buildFastInstantiator = <T extends IClassModelType<Record<string, IAnyType>, any, any>>(model: T): CompiledInstantiator<T> => {
+  return new InstantiatorBuilder(model).build();
+};
+
+type DirectlyAssignableType = SimpleType<any> | IntegerType | LiteralType<any> | DateType;
+const isDirectlyAssignableType = (type: IAnyType): type is DirectlyAssignableType =>
+  type instanceof SimpleType || type instanceof IntegerType || type instanceof LiteralType || type instanceof DateType;
+
+class InstantiatorBuilder<T extends IClassModelType<Record<string, IAnyType>, any, any>> {
+  aliases = new Map<string, string>();
+
+  constructor(readonly model: T) {}
+
+  build(): CompiledInstantiator<T> {
+    const segments: string[] = [];
+
+    for (const [key, type] of Object.entries(this.model.properties)) {
+      if (isDirectlyAssignableType(type)) {
+        segments.push(`
+        // simple type for ${key}
+        instance["${key}"] = ${this.expressionForDirectlyAssignableType(key, type)};
+      `);
+      } else if (type instanceof OptionalType) {
+        segments.push(this.assignmentExpressionForOptionalType(key, type));
+      } else if (type instanceof ReferenceType || type instanceof SafeReferenceType) {
+        segments.push(this.assignmentExpressionForReferenceType(key, type));
+      } else if (type instanceof MapType) {
+        segments.push(this.assignmentExpressionForMapType(key, type));
+      } else {
+        segments.push(`
+          // instantiate fallback for ${key} of type ${type.name}
+          instance["${key}"] = ${this.alias(`model.properties["${key}"]`)}.instantiate(
+            snapshot?.["${key}"], 
+            context, 
+            instance
+          );
+        `);
+      }
+    }
+
+    for (const [key, _metadata] of Object.entries(this.model.volatiles)) {
+      segments.push(`
+      instance["${key}"] = ${this.alias(`model.volatiles["${key}"]`)}.initializer(instance);
+    `);
+    }
+
+    const identifierProp = this.model.mstType.identifierAttribute;
+    if (identifierProp) {
+      segments.push(`
+      const id = instance["${identifierProp}"];
+      instance[$identifier] = id;
+      context.referenceCache.set(id, instance); 
+    `);
+    }
+
+    const innerFunc = `
+      return function Instantiate${this.model.name}(instance, snapshot, context) {
+        ${segments.join("\n")}
+      }
+    `;
+
+    const aliasFuncBody = `
+    const { QuickMap, $identifier } = imports;
+    ${Array.from(this.aliases.entries())
+      .map(([expression, alias]) => `const ${alias} = ${expression};`)
+      .join("\n")}
+
+    ${innerFunc}
+  `;
+
+    // console.log(`function for ${this.model.name}`, "\n\n\n", aliasFuncBody, "\n\n\n");
```

Comment:
🔥 ?

Reply:
I found this was really handy for debugging the generated code so I figured I'd leave it for the next person?